# Exploring bike share data

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description
> This is randomly selected data related to bike share systems for the first six months of 2017 in three large cities: Chicago, New York City, and Washington, DC.

|   **Column**  |                    **Description**                    |
|:-------------:|:-----------------------------------------------------:|
|       X       |                        Trip ID                        |
|   Start.Time  |                Trip start day and time                |
|    End.Time   |                 Trip end day and time                 |
| Trip.Duration |              Duration of trip in seconds              |
| Start.Station |                   Trip start station                  |
|  End.Station  |                    Trip end station                   |
|   User.Type   |          Rider type (Subscriber or Customer)          |
|     Gender    |    Male or Female (Chicago and New York City only)    |
|   Birth.Year  | User's year of birth (Chicago and New York City only) |

### Questions for Analysis

- What is the most common month for bike share use?
- What is the most common start station?
- What are the counts of each user type?

### Import necessary libraries

In [1]:
library(tidyverse)
library(lubridate)

Registered S3 methods overwritten by 'ggplot2':
  method         from 
  [.quosures     rlang
  c.quosures     rlang
  print.quosures rlang
Registered S3 method overwritten by 'rvest':
  method            from
  read_xml.response xml2
-- Attaching packages --------------------------------------- tidyverse 1.2.1 --
v ggplot2 3.1.1       v purrr   0.3.2  
v tibble  2.1.1       v dplyr   0.8.0.1
v tidyr   0.8.3       v stringr 1.4.0  
v readr   1.3.1       v forcats 0.4.0  
-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()

Attaching package: 'lubridate'

The following object is masked from 'package:base':

    date



### Read in csv files

In [2]:
ny = read.csv('./data/new_york_city.csv')
wash = read.csv('./data/washington.csv')
chi = read.csv('./data/chicago.csv')

<a id='wrangling'></a>
## Data Wrangling

<ul>
    <li><a href="#ny">New York City</a></li>
    <li><a href="#chi">Chicago</a></li>
    <li><a href="#wash">Washington, DC</a></li>
</ul>

<a id='ny'></a>
### Explore the NY dataset

In [3]:
# Find number of rows and columns
dim(ny) 

[1] 54770     9

In [4]:
# Check column names
names(ny)

[1] "X"             "Start.Time"    "End.Time"      "Trip.Duration"
[5] "Start.Station" "End.Station"   "User.Type"     "Gender"       
[9] "Birth.Year"

In [5]:
# Check top 6 rows of dataset
head(ny)

X       Start.Time          End.Time            Trip.Duration
1 5688089 2017-06-11 14:55:05 2017-06-11 15:08:21  795         
2 4096714 2017-05-11 15:30:11 2017-05-11 15:41:43  692         
3 2173887 2017-03-29 13:26:26 2017-03-29 13:48:31 1325         
4 3945638 2017-05-08 19:47:18 2017-05-08 19:59:01  703         
5 6208972 2017-06-21 07:49:16 2017-06-21 07:54:46  329         
6 1285652 2017-02-22 18:55:24 2017-02-22 19:12:03  998         
  Start.Station           End.Station            User.Type  Gender Birth.Year
1 Suffolk St & Stanton St W Broadway & Spring St Subscriber Male   1998      
2 Lexington Ave & E 63 St 1 Ave & E 78 St        Subscriber Male   1981      
3 1 Pl & Clinton St       Henry St & Degraw St   Subscriber Male   1987      
4 Barrow St & Hudson St   W 20 St & 8 Ave        Subscriber Female 1986      
5 1 Ave & E 44 St         E 53 St & 3 Ave        Subscriber Male   1992      
6 State St & Smith St     Bond St & Fulton St    Subscriber Male   1986

In [6]:
# Check bottom 6 rows of dataset
tail(ny) 

X       Start.Time          End.Time            Trip.Duration
54765 1293888 2017-02-23 06:14:14 2017-02-23 06:23:32  558         
54766  642855 2017-01-28 16:44:18 2017-01-28 16:48:18  240         
54767 2157959 2017-03-29 06:30:35 2017-03-29 06:32:41  125         
54768 5679624 2017-06-11 12:52:27 2017-06-11 12:58:35  367         
54769 6762960 2017-06-30 07:48:34 2017-06-30 08:17:16 1722         
54770 6078570 2017-06-18 16:20:21 201                   NA         
      Start.Station             End.Station              User.Type  Gender
54765 E 27 St & 1 Ave           E 47 St & Park Ave       Subscriber Male  
54766 W 52 St & 9 Ave           9 Ave & W 45 St          Subscriber Male  
54767 W 84 St & Columbus Ave    W 87 St  & Amsterdam Ave Subscriber Male  
54768 8 Ave & W 33 St           W 45 St & 8 Ave          Subscriber Male  
54769 Cathedral Pkwy & Broadway Broadway & W 51 St       Subscriber Male  
54770                                                                     
      Birth.Year
54765 1984      
54766 1991      
54767 1984      
54768 1954      
54769 1974      
54770   NA

In [7]:
# Check structure of data including datatypes
str(ny)

'data.frame':	54770 obs. of  9 variables:
 $ X            : int  5688089 4096714 2173887 3945638 6208972 1285652 1675753 1692245 2271331 1558339 ...
 $ Start.Time   : Factor w/ 54568 levels "2017-01-01 00:17:01",..: 45448 32799 17316 31589 49688 10220 13390 13509 18111 12449 ...
 $ End.Time     : Factor w/ 54562 levels "201","2017-01-01 00:30:56",..: 45432 32783 17295 31567 49668 10204 13364 13505 18092 12422 ...
 $ Trip.Duration: int  795 692 1325 703 329 998 478 4038 5132 309 ...
 $ Start.Station: Factor w/ 636 levels "","1 Ave & E 16 St",..: 522 406 10 93 5 521 325 309 151 245 ...
 $ End.Station  : Factor w/ 638 levels "","1 Ave & E 16 St",..: 613 8 362 558 269 107 389 110 151 243 ...
 $ User.Type    : Factor w/ 3 levels "","Customer",..: 3 3 3 3 3 3 3 3 2 3 ...
 $ Gender       : Factor w/ 3 levels "","Female","Male": 3 3 3 2 3 3 3 3 1 3 ...
 $ Birth.Year   : num  1998 1981 1987 1986 1992 ...


`Start.Time` and `End.Time` are both factor types and will need to be converted

`End.Time` has a value of 201 instead of a date and time

`User.Type` has 3 levels. We expect Customer and Subscriber and some of the values are missing.

`Gender` has 3 levels. We expect Male and Female and some of the values are missing.

In [8]:
# Create object to store every row that has at least one missing value
ny_missing <- !complete.cases(ny)

# Show all the rows with missing data
ny[ny_missing,]

X       Start.Time          End.Time            Trip.Duration
9     2271331 2017-04-02 08:02:36 2017-04-02 09:28:08 5132         
11    2287178 2017-04-02 14:37:20 2017-04-02 14:56:12 1131         
20       5857 2017-01-01 13:32:39 2017-01-01 13:49:57 1038         
24    2497952 2017-04-08 13:39:48 2017-04-08 14:04:24 1476         
33    3676202 2017-05-02 21:43:28 2017-05-02 22:29:15 2746         
37    1975396 2017-03-22 08:56:43 2017-03-22 09:07:13  630         
39    5630375 2017-06-10 14:03:43 2017-06-10 14:05:00   76         
53    2897347 2017-04-16 15:23:43 2017-04-16 15:44:16 1233         
61    3847598 2017-05-06 15:58:00 2017-05-06 16:31:17 1997         
66    6018157 2017-06-17 08:06:57 2017-06-17 08:35:44 1727         
79    4500991 2017-05-19 13:50:41 2017-05-19 14:02:14  692         
108   6356851 2017-06-23 12:01:36 2017-06-23 12:40:38 2341         
114   1214745 2017-02-20 17:42:58 2017-02-20 18:13:51 1852         
119   4824502 2017-05-26 14:03:32 2017-05-26 14:24:49 1276         
124   2792598 2017-04-14 12:17:13 2017-04-14 12:37:46 1232         
125   3006257 2017-04-18 17:01:54 2017-04-18 17:17:07  912         
138   1669014 2017-03-06 11:57:51 2017-03-06 12:25:57 1686         
140   3712090 2017-05-03 16:20:03 2017-05-03 17:35:56 4552         
141   5529352 2017-06-08 19:14:41 2017-06-08 19:46:30 1909         
142   6467971 2017-06-25 14:33:29 2017-06-25 14:58:08 1479         
169     13703 2017-01-01 19:12:56 2017-01-01 19:31:44 1127         
184   6441021 2017-06-24 22:21:47 2017-06-24 22:41:18 1171         
192   2486890 2017-04-08 09:02:24 2017-04-08 09:21:25 1141         
198   2320738 2017-04-03 09:33:57 2017-04-03 09:55:43 1306         
204   4218781 2017-05-14 19:03:42 2017-05-14 19:21:48 1085         
208   2744300 2017-04-13 13:27:15 2017-04-13 13:52:24 1509         
211   5116172 2017-06-01 14:49:27 2017-06-01 15:18:07 1719         
217   1157420 2017-02-19 10:33:36 2017-02-19 11:03:57 1821         
225   6451583 2017-06-25 10:17:15 2017-06-25 10:40:24 1388         
226   4519233 2017-05-19 17:58:47 2017-05-19 18:15:59 1032         
...   ...     ...                 ...                 ...          
54430  758074 2017-02-01 19:55:49 2017-02-01 20:01:31    342       
54435 2879094 2017-04-16 11:10:07 2017-04-16 12:18:56   4128       
54436 4589348 2017-05-21 09:52:04 2017-05-23 15:39:58 193674       
54437 2752167 2017-04-13 16:08:26 2017-04-13 16:24:49    983       
54467 6539497 2017-06-26 17:51:55 2017-06-26 18:15:20   1405       
54473 3481495 2017-04-29 13:24:35 2017-04-29 13:35:22    646       
54484 3474771 2017-04-29 11:59:46 2017-04-29 12:21:35   1309       
54497 6259774 2017-06-21 19:51:19 2017-06-21 20:06:50    930       
54503 2055901 2017-03-24 20:03:30 2017-03-24 20:11:02    452       
54508 2891743 2017-04-16 14:10:58 2017-04-16 14:33:07   1329       
54512 4929181 2017-05-28 15:55:45 2017-05-28 17:19:04   4999       
54525 3728096 2017-05-03 19:05:05 2017-05-03 19:27:30   1345       
54535 3509312 2017-04-29 19:25:34 2017-04-29 19:55:30   1796       
54554 5449439 2017-06-07 16:50:40 2017-06-07 17:18:41   1681       
54558 4884032 2017-05-27 16:14:01 2017-05-27 16:39:59   1557       
54562 5239122 2017-06-03 13:36:22 2017-06-03 13:48:30    727       
54566 3353340 2017-04-27 08:30:40 2017-04-27 08:44:13    812       
54567 4445723 2017-05-18 16:39:44 2017-05-18 17:03:00   1396       
54568 5047988 2017-05-31 14:03:24 2017-05-31 14:19:19    954       
54576 5257523 2017-06-03 17:23:51 2017-06-03 18:03:32   2380       
54605 6307929 2017-06-22 16:51:09 2017-06-22 17:11:58   1248       
54652 5857305 2017-06-14 14:11:43 2017-06-14 14:16:35    292       
54654 1398901 2017-02-25 13:48:32 2017-02-25 14:48:55   3623       
54668 4427607 2017-05-18 10:02:53 2017-05-18 10:50:06   2832       
54685 3225885 2017-04-23 17:06:04 2017-04-23 17:27:38   1293       
54696  494070 2017-01-22 19:27:48 2017-01-22 19:35:07    438       
54703 6368841 2017-06-23 15:56:44 2017-06-23 16:14:42   1077  

Most of the missing values appears to be coming from `Gender` and `Birth.Year`.

`X` ID of 6078570 is missing values or has incorrect values in almost all the columns. This record will need to be removed completely.

In [9]:
# Count duplicated rows
sum(duplicated(ny))

[1] 0

<a id='chi'></a>
### Explore the Chicago dataset

In [10]:
# Find number of rows and columns
dim(chi) 

[1] 8630    9

In [11]:
# Check column names
names(chi)

[1] "X"             "Start.Time"    "End.Time"      "Trip.Duration"
[5] "Start.Station" "End.Station"   "User.Type"     "Gender"       
[9] "Birth.Year"

In [12]:
# Check top 6 rows of dataset
head(chi)

X       Start.Time          End.Time            Trip.Duration
1 1423854 2017-06-23 15:09:32 2017-06-23 15:14:53  321         
2  955915 2017-05-25 18:19:03 2017-05-25 18:45:53 1610         
3    9031 2017-01-04 08:27:49 2017-01-04 08:34:45  416         
4  304487 2017-03-06 13:49:38 2017-03-06 13:55:28  350         
5   45207 2017-01-17 14:53:07 2017-01-17 15:02:01  534         
6 1473887 2017-06-26 09:01:20 2017-06-26 09:11:06  586         
  Start.Station                 End.Station                  User.Type  Gender
1 Wood St & Hubbard St          Damen Ave & Chicago Ave      Subscriber Male  
2 Theater on the Lake           Sheffield Ave & Waveland Ave Subscriber Female
3 May St & Taylor St            Wood St & Taylor St          Subscriber Male  
4 Christiana Ave & Lawrence Ave St. Louis Ave & Balmoral Ave Subscriber Male  
5 Clark St & Randolph St        Desplaines St & Jackson Blvd Subscriber Male  
6 Clinton St & Washington Blvd  Canal St & Taylor St         Subscriber Male  
  Birth.Year
1 1992      
2 1992      
3 1981      
4 1986      
5 1975      
6 1990

In [13]:
# Check bottom 6 rows of dataset
tail(chi)

X       Start.Time          End.Time            Trip.Duration
8625  397518 2017-03-24 16:52:16 2017-03-24 16:57:57  341         
8626  879494 2017-05-18 05:06:50 2017-05-18 05:22:10  920         
8627  360389 2017-03-19 07:21:29 2017-03-19 07:27:18  349         
8628  858496 2017-05-16 17:03:24 2017-05-16 17:31:12 1668         
8629  777620 2017-05-10 08:53:03 2017-05-10 08:54:32   89         
8630 1230561 2017-06-11 14:52:13 2017-06-11 15:42:33 3020         
     Start.Station                End.Station                  User.Type 
8625 Southport Ave & Waveland Ave Southport Ave & Waveland Ave Subscriber
8626 Artesian Ave & Hubbard St    Wacker Dr & Washington St    Subscriber
8627 Wabash Ave & Roosevelt Rd    Wells St & Polk St           Subscriber
8628 Ashland Ave & Harrison St    Wells St & Concord Ln        Subscriber
8629 Western Ave & Leland Ave     Western Ave & Leland Ave     Subscriber
8630 Waba                                                                
     Gender Birth.Year
8625 Male   1990      
8626 Male   1959      
8627 Male   1987      
8628 Male   1963      
8629 Male   1977      
8630          NA

In [14]:
# Check structure of data including datatypes
str(chi)

'data.frame':	8630 obs. of  9 variables:
 $ X            : int  1423854 955915 9031 304487 45207 1473887 961916 65924 606841 135470 ...
 $ Start.Time   : Factor w/ 8624 levels "2017-01-01 00:40:14",..: 7876 5303 73 1721 267 8173 5347 368 3376 795 ...
 $ End.Time     : Factor w/ 8625 levels "2017-01-01 00:46:32",..: 7876 5303 73 1722 267 8173 5346 368 3376 796 ...
 $ Trip.Duration: int  321 1610 416 350 534 586 281 723 689 493 ...
 $ Start.Station: Factor w/ 472 levels "2112 W Peterson Ave",..: 468 424 291 80 103 119 22 255 374 420 ...
 $ End.Station  : Factor w/ 471 levels "","2112 W Peterson Ave",..: 132 381 469 409 151 70 467 251 200 118 ...
 $ User.Type    : Factor w/ 3 levels "","Customer",..: 3 3 3 3 3 3 3 2 3 3 ...
 $ Gender       : Factor w/ 3 levels "","Female","Male": 3 2 3 3 3 3 2 1 3 3 ...
 $ Birth.Year   : num  1992 1992 1981 1986 1975 ...


`Start.Time` and `End.Time` are both factor types and will need to be converted

`End.Station` has a missing value and will need to be investigated when cleaning.

`User.Type` has 3 levels. We expect Customer and Subscriber and some of the values are missing.

`Gender` has 3 levels. We expect Male and Female and some of the values are missing.

In [15]:
# Create object to store every row that has at least one missing value
chi_missing <- !complete.cases(chi)

# Show all the rows with missing data
chi[chi_missing,]

X       Start.Time          End.Time            Trip.Duration
8      65924 2017-01-21 14:28:38 2017-01-21 14:40:41   723        
20    475456 2017-04-08 11:37:55 2017-04-08 11:51:55   840        
32   1539334 2017-06-30 10:56:50 2017-06-30 11:40:20  2610        
36    243879 2017-02-22 15:33:56 2017-02-22 15:54:07  1211        
39    720062 2017-05-03 16:27:08 2017-05-03 16:45:15  1087        
41   1314009 2017-06-16 19:34:44 2017-06-16 20:16:23  2499        
45   1372709 2017-06-20 16:14:15 2017-06-20 16:42:26  1691        
53    157790 2017-02-11 15:11:34 2017-02-11 16:30:04  4710        
57   1526760 2017-06-29 13:50:47 2017-06-29 14:10:04  1157        
62   1539175 2017-06-30 10:44:24 2017-06-30 11:11:03  1599        
69   1341523 2017-06-18 16:17:10 2017-06-18 16:53:55  2205        
77   1541842 2017-06-30 13:48:27 2017-06-30 14:13:56  1529        
80   1462466 2017-06-25 15:09:23 2017-06-25 15:30:45  1282        
86    635646 2017-04-23 14:28:35 2017-04-23 15:00:52  1937        
87    200941 2017-02-18 13:35:55 2017-02-18 14:13:54  2279        
96   1550326 2017-06-30 21:03:10 2017-06-30 21:27:49  1479        
99    919854 2017-05-22 11:44:34 2017-05-22 12:08:13  1419        
103   519940 2017-04-12 16:31:59 2017-04-12 17:03:43  1904        
106  1245581 2017-06-12 13:36:58 2017-06-12 15:04:24  5246        
107   748995 2017-05-06 20:28:25 2017-05-06 20:46:42  1097        
119   264023 2017-02-26 15:54:07 2017-02-26 16:49:43  3336        
126  1473181 2017-06-26 08:30:03 2017-06-26 10:38:24  7701        
137   533191 2017-04-14 12:00:44 2017-04-14 12:22:51  1327        
151  1314112 2017-06-16 19:41:46 2017-06-16 20:08:47  1621        
152   632189 2017-04-23 11:37:03 2017-04-23 12:53:12  4569        
153   543142 2017-04-15 11:38:34 2017-04-15 14:49:56 11482        
159  1404482 2017-06-22 11:04:37 2017-06-22 11:21:04   987        
164  1181708 2017-06-08 19:24:41 2017-06-08 20:04:22  2381        
170   626308 2017-04-22 15:22:07 2017-04-22 15:28:06   359        
177   854604 2017-05-16 12:48:26 2017-05-16 13:03:21   895        
...  ...     ...                 ...                 ...          
8480  985688 2017-05-28 09:02:41 2017-05-28 10:27:46 5105         
8483  564200 2017-04-16 16:42:31 2017-04-16 18:24:02 6091         
8485  488429 2017-04-09 11:59:13 2017-04-09 12:54:36 3323         
8487  394250 2017-03-24 13:43:10 2017-03-24 14:07:36 1466         
8508  555458 2017-04-15 19:16:50 2017-04-15 19:30:04  794         
8510  329329 2017-03-10 06:45:52 2017-03-10 06:55:03  551         
8518 1066714 2017-06-02 09:54:47 2017-06-02 10:16:41 1314         
8533  520452 2017-04-12 16:58:38 2017-04-12 19:06:27 7669         
8541 1321809 2017-06-17 12:48:12 2017-06-17 13:58:36 4224         
8544 1457535 2017-06-25 11:56:16 2017-06-25 13:15:56 4780         
8546  406515 2017-03-27 09:17:46 2017-03-27 09:45:09 1643         
8551 1255988 2017-06-13 06:30:31 2017-06-13 06:49:38 1147         
8554  813915 2017-05-13 13:52:55 2017-05-13 14:05:03  728         
8559  672587 2017-04-26 10:53:30 2017-04-26 11:17:45 1455         
8560  889464 2017-05-18 17:25:54 2017-05-18 18:40:30 4476         
8561 1011416 2017-05-29 17:27:52 2017-05-29 18:25:51 3479         
8562  979318 2017-05-27 16:59:10 2017-05-27 17:20:00 1250         
8564  797904 2017-05-12 10:27:01 2017-05-12 11:05:02 2281         
8565 1516988 2017-06-28 18:31:11 2017-06-28 18:53:55 1364         
8590 1042403 2017-05-31 18:14:18 2017-05-31 18:20:08  350         
8593 1009880 2017-05-29 16:30:53 2017-05-29 16:58:38 1665         
8596  839454 2017-05-15 13:12:32 2017-05-15 13:32:01 1169         
8602 1371090 2017-06-20 14:05:16 2017-06-20 14:17:43  747         
8603  550447 2017-04-15 15:45:09 2017-04-15 16:11:20 1571         
8604  243811 2017-02-22 15:28:34 2017-02-22 15:38:38  604         
8606 1406209 2017-06-22 13:15:06 2017-06-22 14:10:15 3309         
8609 1464943 2017-06-25 16:43:18 2017-06-25 16:59:02  944         
8620 1397375 2017-06-21 20:06:34 2017-06-21 20:25:10

Most of the missing values appears to be coming from `Gender` and `Birth.Year`.

`X` ID of 1230561 is missing values. This record will need to be removed completely.

In [16]:
# Count duplicated rows
sum(duplicated(chi))

[1] 0

<a id='wash'></a>
### Explore the Washington, DC dataset

In [17]:
# Find number of rows and columns
dim(wash) 

[1] 89051     7

In [18]:
# Check column names
names(wash)

[1] "X"             "Start.Time"    "End.Time"      "Trip.Duration"
[5] "Start.Station" "End.Station"   "User.Type"

In [19]:
# Check top 6 rows of dataset
head(wash)

X       Start.Time          End.Time            Trip.Duration
1 1621326 2017-06-21 08:36:34 2017-06-21 08:44:43  489.066     
2  482740 2017-03-11 10:40:00 2017-03-11 10:46:00  402.549     
3 1330037 2017-05-30 01:02:59 2017-05-30 01:13:37  637.251     
4  665458 2017-04-02 07:48:35 2017-04-02 08:19:03 1827.341     
5 1481135 2017-06-10 08:36:28 2017-06-10 09:02:17 1549.427     
6 1148202 2017-05-14 07:18:18 2017-05-14 07:24:56  398.000     
  Start.Station                              
1 14th & Belmont St NW                       
2 Yuma St & Tenley Circle NW                 
3 17th St & Massachusetts Ave NW             
4 Constitution Ave & 2nd St NW/DOL           
5 Henry Bacon Dr & Lincoln Memorial Circle NW
6 1st & K St SE                              
  End.Station                                         User.Type 
1 15th & K St NW                                      Subscriber
2 Connecticut Ave & Yuma St NW                        Subscriber
3 5th & K St NW                                       Subscriber
4 M St & Pennsylvania Ave NW                          Customer  
5 Maine Ave & 7th St SW                               Subscriber
6 Eastern Market Metro / Pennsylvania Ave & 7th St SE Subscriber

In [20]:
# Check top 6 rows of dataset
tail(wash)

X       Start.Time          End.Time            Trip.Duration
89046 1484340 2017-06-10 10:58:09 2017-06-10 11:25:58 1669.700     
89047  555788 2017-03-22 18:46:00 2017-03-22 19:04:00 1082.789     
89048  739004 2017-04-09 04:00:22 2017-04-09 04:09:54  571.879     
89049 1214907 2017-05-19 09:00:53 2017-05-19 09:07:38  404.152     
89050 1419806 2017-06-06 04:27:33 2017-06-06 04:49:59 1345.911     
89051     132                                               NA     
      Start.Station            End.Station                     User.Type 
89046 M St & New Jersey Ave SE 4th St & Madison Dr NW          Customer  
89047 8th & H St NW            21st & I St NW                  Subscriber
89048 Eckington Pl & Q St NE   Columbus Circle / Union Station Subscriber
89049 1st & M St NE            1st & Rhode Island Ave NW       Subscriber
89050 10th & Florida Ave NW    Georgetown Harbor / 30th St NW  Customer  
89051

In [21]:
# Check structure of data including datatypes
str(wash)

'data.frame':	89051 obs. of  7 variables:
 $ X            : int  1621326 482740 1330037 665458 1481135 1148202 1594275 1601832 574182 327058 ...
 $ Start.Time   : Factor w/ 81223 levels "","2017-01-01 00:11:00",..: 74753 19510 59964 26708 67716 50891 73381 73775 23142 13333 ...
 $ End.Time     : Factor w/ 81217 levels "","2017-01-01 00:14:00",..: 74744 19473 59981 26732 67753 50918 73397 73775 23114 13350 ...
 $ Trip.Duration: num  489 403 637 1827 1549 ...
 $ Start.Station: Factor w/ 478 levels "","10th & E St NW",..: 27 478 66 221 278 84 368 82 71 60 ...
 $ End.Station  : Factor w/ 479 levels "","10th & E St NW",..: 47 219 144 312 315 239 162 376 51 308 ...
 $ User.Type    : Factor w/ 3 levels "","Customer",..: 3 3 3 2 3 3 3 3 3 3 ...


`Start.Time` and `End.Time` are both factor types and will need to be converted

`Start.Station` and `End.Station` have missing values and will need to be investigated when cleaning.

`Trip.Duration` is a num instead of int in the `Chicago` and `New York City` datasets.

`End.Station` has a missing value and will need to be investigated when cleaning.

`User.Type` has 3 levels. We expect Customer and Subscriber and some of the values are missing.

In [22]:
# Create object to store every row that has at least one missing value
wash_missing <- !complete.cases(wash)

# Show all the rows with missing data
wash[wash_missing,]

X   Start.Time End.Time Trip.Duration Start.Station End.Station User.Type
89051 132                     NA

`X` ID of 132 is missing values. This record will need to be removed completely.

In [23]:
# Count duplicated rows
sum(duplicated(wash))

[1] 0

In [24]:
sum(duplicated(wash$User.Type))

[1] 89048

### Data Cleaning

To help answer the analysis questions each dataset will need to be cleaned. The following steps will be taken to clean the data:

- Remove columns that are not needed to answer analysis questions
- Update and transform data in columns for easier understanding and consistency
- Remove records with incorrect or missing values
- Add new columns to assist with answering questions
- Merge each new clean dataset into a single dataset

In [25]:
# Make copies of all the datasets before cleaning
# Creates a copy with a different memory address
# https://rdrr.io/cran/data.table/man/copy.html

ny_clean <- data.table::copy(ny)
chi_clean <- data.table::copy(chi)
wash_clean <- data.table::copy(wash)

#### Clean New York City Dataset

In [26]:
names(ny_clean)

[1] "X"             "Start.Time"    "End.Time"      "Trip.Duration"
[5] "Start.Station" "End.Station"   "User.Type"     "Gender"       
[9] "Birth.Year"

In [27]:
# Remove Gender and Birth.Year columns
ny_clean <- ny_clean %>%
              select(X, Start.Time, End.Time, Trip.Duration, Start.Station, End.Station, User.Type)

In [28]:
# Check that columns have been removed
names(ny_clean)

[1] "X"             "Start.Time"    "End.Time"      "Trip.Duration"
[5] "Start.Station" "End.Station"   "User.Type"

In [29]:
# Add column for city name for merging into single dataframe
ny_clean <- ny_clean %>%
              mutate(city = 'NYC') %>%
              select(X, Start.Time, End.Time, Trip.Duration, Start.Station, End.Station, User.Type, city)

In [30]:
# Check that city column has been added
sample_n(ny_clean, 5)

X       Start.Time          End.Time            Trip.Duration
1 2378101 2017-04-04 19:34:51 2017-04-04 19:48:50  838         
2 2393600 2017-04-05 09:13:00 2017-04-05 09:24:42  701         
3 1565900 2017-03-02 08:38:55 2017-03-02 08:51:48  773         
4 6380845 2017-06-23 18:14:42 2017-06-23 18:19:15  273         
5  447760 2017-01-20 20:37:59 2017-01-20 21:03:42 1543         
  Start.Station                 End.Station           User.Type  city
1 Christopher St & Greenwich St Fulton St & Broadway  Subscriber NYC 
2 W 13 St & 7 Ave               Grand St & Greene St  Subscriber NYC 
3 E 2 St & 2 Ave E              W 13 St & Hudson St   Subscriber NYC 
4 W Broadway & Spring St        Barrow St & Hudson St Subscriber NYC 
5 W 39 St & 9 Ave               Liberty St & Broadway Subscriber NYC

In [31]:
# Add start_date column with date from Start.Time
ny_clean$start_date <- as.Date(ny_clean$Start.Time)

In [32]:
# Check that new start_date column is created and only has dates
sample_n(ny_clean, 5)

X       Start.Time          End.Time            Trip.Duration
1 2196909 2017-03-29 21:06:39 2017-03-29 21:13:40  420         
2 3741722 2017-05-04 07:24:33 2017-05-04 07:45:59 1285         
3 3152607 2017-04-21 20:17:46 2017-04-21 20:28:53  667         
4 2643990 2017-04-11 15:36:19 2017-04-11 17:14:01 5861         
5 5828768 2017-06-13 21:05:39 2017-06-13 21:29:17 1418         
  Start.Station          End.Station            User.Type  city start_date
1 W 27 St & 10 Ave       Broadway & W 32 St     Subscriber NYC  2017-03-29
2 W 64 St & West End Ave E 55 St & 2 Ave        Subscriber NYC  2017-05-04
3 9 Ave & W 18 St        W 45 St & 8 Ave        Subscriber NYC  2017-04-21
4 Broadway & W 51 St     E 58 St & 3 Ave        Customer   NYC  2017-04-11
5 Fulton St & William St Bayard St & Leonard St Subscriber NYC  2017-06-13

In [33]:
# Check data types
str(ny_clean)

'data.frame':	54770 obs. of  9 variables:
 $ X            : int  5688089 4096714 2173887 3945638 6208972 1285652 1675753 1692245 2271331 1558339 ...
 $ Start.Time   : Factor w/ 54568 levels "2017-01-01 00:17:01",..: 45448 32799 17316 31589 49688 10220 13390 13509 18111 12449 ...
 $ End.Time     : Factor w/ 54562 levels "201","2017-01-01 00:30:56",..: 45432 32783 17295 31567 49668 10204 13364 13505 18092 12422 ...
 $ Trip.Duration: int  795 692 1325 703 329 998 478 4038 5132 309 ...
 $ Start.Station: Factor w/ 636 levels "","1 Ave & E 16 St",..: 522 406 10 93 5 521 325 309 151 245 ...
 $ End.Station  : Factor w/ 638 levels "","1 Ave & E 16 St",..: 613 8 362 558 269 107 389 110 151 243 ...
 $ User.Type    : Factor w/ 3 levels "","Customer",..: 3 3 3 3 3 3 3 3 2 3 ...
 $ city         : chr  "NYC" "NYC" "NYC" "NYC" ...
 $ start_date   : Date, format: "2017-06-11" "2017-05-11" ...


In [34]:
# Add end_date column with date from End.Time
ny_clean$end_date <- as.Date(ny_clean$End.Time)

In [35]:
# Check that new end_date column is created and only has dates
sample_n(ny_clean, 5)

X       Start.Time          End.Time            Trip.Duration
1 5524074 2017-06-08 18:20:43 2017-06-08 18:30:33  590         
2 1837213 2017-03-11 16:43:49 2017-03-11 16:45:39  110         
3  821176 2017-02-03 18:38:07 2017-02-03 18:41:58  230         
4 6736764 2017-06-29 17:49:19 2017-06-29 18:04:10  891         
5 2585269 2017-04-10 14:00:51 2017-04-10 14:21:26 1234         
  Start.Station               End.Station           User.Type  city start_date
1 W 27 St & 7 Ave             W 47 St & 10 Ave      Subscriber NYC  2017-06-08
2 9 Ave & W 45 St             W 43 St & 10 Ave      Subscriber NYC  2017-03-11
3 Lewis Ave & Decatur St      Fulton St & Utica Ave Subscriber NYC  2017-02-03
4 8 Ave & W 33 St             Carmine St & 6 Ave    Subscriber NYC  2017-06-29
5 Pier 40 - Hudson River Park E 17 St & Broadway    Customer   NYC  2017-04-10
  end_date  
1 2017-06-08
2 2017-03-11
3 2017-02-03
4 2017-06-29
5 2017-04-10

In [36]:
# Check data types
str(ny_clean)

'data.frame':	54770 obs. of  10 variables:
 $ X            : int  5688089 4096714 2173887 3945638 6208972 1285652 1675753 1692245 2271331 1558339 ...
 $ Start.Time   : Factor w/ 54568 levels "2017-01-01 00:17:01",..: 45448 32799 17316 31589 49688 10220 13390 13509 18111 12449 ...
 $ End.Time     : Factor w/ 54562 levels "201","2017-01-01 00:30:56",..: 45432 32783 17295 31567 49668 10204 13364 13505 18092 12422 ...
 $ Trip.Duration: int  795 692 1325 703 329 998 478 4038 5132 309 ...
 $ Start.Station: Factor w/ 636 levels "","1 Ave & E 16 St",..: 522 406 10 93 5 521 325 309 151 245 ...
 $ End.Station  : Factor w/ 638 levels "","1 Ave & E 16 St",..: 613 8 362 558 269 107 389 110 151 243 ...
 $ User.Type    : Factor w/ 3 levels "","Customer",..: 3 3 3 3 3 3 3 3 2 3 ...
 $ city         : chr  "NYC" "NYC" "NYC" "NYC" ...
 $ start_date   : Date, format: "2017-06-11" "2017-05-11" ...
 $ end_date     : Date, format: "2017-06-11" "2017-05-11" ...


In [37]:
# Check unique values for end_date
unique(ny_clean$end_date)

[1] "2017-06-11" "2017-05-11" "2017-03-29" "2017-05-08" "2017-06-21"
  [6] "2017-02-22" "2017-03-06" "2017-03-07" "2017-04-02" "2017-03-01"
 [11] "2017-04-13" "2017-04-28" "2017-02-13" "2017-02-28" "2017-01-11"
 [16] "2017-06-23" "2017-05-24" "2017-01-01" "2017-02-18" "2017-04-27"
 [21] "2017-03-09" "2017-04-08" "2017-04-16" "2017-04-21" "2017-04-17"
 [26] "2017-03-25" "2017-04-30" "2017-06-06" "2017-03-23" "2017-05-02"
 [31] "2017-02-25" "2017-06-22" "2017-02-08" "2017-03-22" "2017-01-28"
 [36] "2017-06-10" "2017-06-08" "2017-05-07" "2017-04-10" "2017-06-24"
 [41] "2017-03-20" "2017-05-30" "2017-05-19" "2017-06-28" "2017-06-05"
 [46] "2017-01-26" "2017-06-16" "2017-05-01" "2017-06-07" "2017-05-20"
 [51] "2017-05-27" "2017-05-10" "2017-05-06" "2017-05-21" "2017-04-25"
 [56] "2017-06-29" "2017-06-17" "2017-01-05" "2017-04-15" "2017-01-17"
 [61] "2017-01-20" "2017-06-15" "2017-04-24" "2017-05-15" "2017-05-16"
 [66] "2017-02-24" "2017-04-29" "2017-06-18" "2017-06-14" "2017-04-05"
 [71] "2017-06-13" "2017-03-21" "2017-04-18" "2017-05-14" "2017-04-12"
 [76] "2017-05-17" "2017-03-10" "2017-01-19" "2017-05-23" "2017-04-20"
 [81] "2017-06-20" "2017-02-20" "2017-04-19" "2017-03-13" "2017-05-26"
 [86] "2017-06-02" "2017-03-02" "2017-01-18" "2017-03-08" "2017-04-14"
 [91] "2017-04-09" "2017-02-04" "2017-05-04" "2017-04-23" "2017-02-23"
 [96] "2017-05-03" "2017-06-25" "2017-02-02" "2017-04-06" "2017-01-04"
[101] "2017-03-28" "2017-01-12" "2017-03-17" "2017-04-03" "2017-02-11"
[106] "2017-06-04" "2017-01-30" "2017-06-09" "2017-04-07" "2017-01-29"
[111] "2017-05-05" "2017-03-05" "2017-04-26" "2017-03-27" "2017-06-01"
[116] "2017-01-13" "2017-02-19" "2017-05-22" "2017-06-12" "2017-02-21"
[121] "2017-01-22" "2017-02-14" "2017-06-30" "2017-03-24" "2017-06-27"
[126] "2017-03-12" "2017-05-18" "2017-02-03" "2017-04-11" "2017-01-23"
[131] "2017-01-10" "2017-01-16" "2017-06-03" "2017-05-31" "2017-03-03"
[136] "2017-03-30" "2017-06-19" "2017-06-26" "2017-01-24" "2017-01-25"
[141] "2017-02-27" "2017-05-12" "2017-05-09" "2017-03-11" "2017-04-01"
[146] "2017-02-26" "2017-05-25" "2017-03-19" "2017-04-22" "2017-01-06"
[151] "2017-02-06" "2017-02-07" "2017-01-07" "2017-01-03" "2017-02-16"
[156] "2017-03-26" "2017-02-01" "2017-03-04" "2017-04-04" "2017-01-09"
[161] "2017-01-21" "2017-02-17" "2017-01-27" "2017-05-28" "2017-01-31"
[166] "2017-01-02" "2017-07-01" "2017-01-15" "2017-02-15" "2017-05-29"
[171] "2017-01-14" "2017-02-05" "2017-03-31" "2017-02-10" "2017-01-08"
[176] "2017-02-12" "2017-03-18" "2017-05-13" NA

In [38]:
# Keep only records that have a value in the end date field
ny_clean <- ny_clean %>%
              select(everything()) %>%
              filter(complete.cases(ny_clean$end_date))

In [39]:
# Check that no values are empty
unique(ny_clean$end_date)

[1] "2017-06-11" "2017-05-11" "2017-03-29" "2017-05-08" "2017-06-21"
  [6] "2017-02-22" "2017-03-06" "2017-03-07" "2017-04-02" "2017-03-01"
 [11] "2017-04-13" "2017-04-28" "2017-02-13" "2017-02-28" "2017-01-11"
 [16] "2017-06-23" "2017-05-24" "2017-01-01" "2017-02-18" "2017-04-27"
 [21] "2017-03-09" "2017-04-08" "2017-04-16" "2017-04-21" "2017-04-17"
 [26] "2017-03-25" "2017-04-30" "2017-06-06" "2017-03-23" "2017-05-02"
 [31] "2017-02-25" "2017-06-22" "2017-02-08" "2017-03-22" "2017-01-28"
 [36] "2017-06-10" "2017-06-08" "2017-05-07" "2017-04-10" "2017-06-24"
 [41] "2017-03-20" "2017-05-30" "2017-05-19" "2017-06-28" "2017-06-05"
 [46] "2017-01-26" "2017-06-16" "2017-05-01" "2017-06-07" "2017-05-20"
 [51] "2017-05-27" "2017-05-10" "2017-05-06" "2017-05-21" "2017-04-25"
 [56] "2017-06-29" "2017-06-17" "2017-01-05" "2017-04-15" "2017-01-17"
 [61] "2017-01-20" "2017-06-15" "2017-04-24" "2017-05-15" "2017-05-16"
 [66] "2017-02-24" "2017-04-29" "2017-06-18" "2017-06-14" "2017-04-05"
 [71] "2017-06-13" "2017-03-21" "2017-04-18" "2017-05-14" "2017-04-12"
 [76] "2017-05-17" "2017-03-10" "2017-01-19" "2017-05-23" "2017-04-20"
 [81] "2017-06-20" "2017-02-20" "2017-04-19" "2017-03-13" "2017-05-26"
 [86] "2017-06-02" "2017-03-02" "2017-01-18" "2017-03-08" "2017-04-14"
 [91] "2017-04-09" "2017-02-04" "2017-05-04" "2017-04-23" "2017-02-23"
 [96] "2017-05-03" "2017-06-25" "2017-02-02" "2017-04-06" "2017-01-04"
[101] "2017-03-28" "2017-01-12" "2017-03-17" "2017-04-03" "2017-02-11"
[106] "2017-06-04" "2017-01-30" "2017-06-09" "2017-04-07" "2017-01-29"
[111] "2017-05-05" "2017-03-05" "2017-04-26" "2017-03-27" "2017-06-01"
[116] "2017-01-13" "2017-02-19" "2017-05-22" "2017-06-12" "2017-02-21"
[121] "2017-01-22" "2017-02-14" "2017-06-30" "2017-03-24" "2017-06-27"
[126] "2017-03-12" "2017-05-18" "2017-02-03" "2017-04-11" "2017-01-23"
[131] "2017-01-10" "2017-01-16" "2017-06-03" "2017-05-31" "2017-03-03"
[136] "2017-03-30" "2017-06-19" "2017-06-26" "2017-01-24" "2017-01-25"
[141] "2017-02-27" "2017-05-12" "2017-05-09" "2017-03-11" "2017-04-01"
[146] "2017-02-26" "2017-05-25" "2017-03-19" "2017-04-22" "2017-01-06"
[151] "2017-02-06" "2017-02-07" "2017-01-07" "2017-01-03" "2017-02-16"
[156] "2017-03-26" "2017-02-01" "2017-03-04" "2017-04-04" "2017-01-09"
[161] "2017-01-21" "2017-02-17" "2017-01-27" "2017-05-28" "2017-01-31"
[166] "2017-01-02" "2017-07-01" "2017-01-15" "2017-02-15" "2017-05-29"
[171] "2017-01-14" "2017-02-05" "2017-03-31" "2017-02-10" "2017-01-08"
[176] "2017-02-12" "2017-03-18" "2017-05-13"

In [40]:
# Make sure there are no rows with missing values
ny_missing <- !complete.cases(ny_clean)
ny_clean[ny_missing,]

X Start.Time End.Time Trip.Duration Start.Station End.Station User.Type
     city start_date end_date

In [41]:
table(ny_clean$User.Type)


             Customer Subscriber 
       118       5558      49093 

#### Clean Chicago Dataset

#### Clean Washington Dataset

#### Merge datasets

<a id='eda'></a>
## Exploratory Data Analysis

### What is the most common month for bike share use?

### What is the most common start station?

### What are the counts of each user type?

<a id='conclusions'></a>
## Conclusions

#### Most common month


#### Most common start station


#### Counts of each user type



### Limitation